File: src-tauri/src/theme.rs
use serde::Serialize;
use std::collections::HashMap;
use std::fs;
use std::path::PathBuf;

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct Theme {
    pub background: String,
    pub foreground: String,
    pub cursor: String,
    pub cursor_accent: String,
    pub selection_background: String,
    pub selection_foreground: String,
    pub black: String,
    pub red: String,
    pub green: String,
    pub yellow: String,
    pub blue: String,
    pub magenta: String,
    pub cyan: String,
    pub white: String,
    pub bright_black: String,
    pub bright_red: String,
    pub bright_green: String,
    pub bright_yellow: String,
    pub bright_blue: String,
    pub bright_magenta: String,
    pub bright_cyan: String,
    pub bright_white: String,
}

impl Default for Theme {
    fn default() -> Self {
        // Catppuccin Mocha fallback
        Self {
            background: "#1e1e2e".into(),
            foreground: "#cdd6f4".into(),
            cursor: "#f5e0dc".into(),
            cursor_accent: "#1e1e2e".into(),
            selection_background: "#585b70".into(),
            selection_foreground: "#cdd6f4".into(),
            black: "#45475a".into(),
            red: "#f38ba8".into(),
            green: "#a6e3a1".into(),
            yellow: "#f9e2af".into(),
            blue: "#89b4fa".into(),
            magenta: "#f5c2e7".into(),
            cyan: "#94e2d5".into(),
            white: "#a6adc8".into(),
            bright_black: "#585b70".into(),
            bright_red: "#f37799".into(),
            bright_green: "#89d88b".into(),
            bright_yellow: "#ebd391".into(),
            bright_blue: "#74a8fc".into(),
            bright_magenta: "#f2aede".into(),
            bright_cyan: "#6bd7ca".into(),
            bright_white: "#bac2de".into(),
        }
    }
}

fn ghostty_config_path() -> Option<PathBuf> {
    // macOS: ~/Library/Application Support/com.mitchellh.ghostty/config
    if let Some(data) = dirs::data_dir() {
        let p = data.join("com.mitchellh.ghostty/config");
        if p.exists() {
            return Some(p);
        }
    }
    // XDG: ~/.config/ghostty/config
    if let Some(config) = dirs::config_dir() {
        let p = config.join("ghostty/config");
        if p.exists() {
            return Some(p);
        }
    }
    None
}

fn ghostty_themes_dir() -> Option<PathBuf> {
    // Check app bundle first (macOS)
    let app_themes =
        PathBuf::from("/Applications/Ghostty.app/Contents/Resources/ghostty/themes");
    if app_themes.is_dir() {
        return Some(app_themes);
    }
    // XDG data dirs
    if let Some(data) = dirs::data_dir() {
        let p = data.join("ghostty/themes");
        if p.is_dir() {
            return Some(p);
        }
    }
    None
}

fn load_theme_from_config(config: &HashMap<String, String>) -> Theme {
    let mut theme = Theme::default();

    if let Some(bg) = config.get("background") {
        theme.background = bg.clone();
        theme.cursor_accent = bg.clone();
    }
    if let Some(fg) = config.get("foreground") {
        theme.foreground = fg.clone();
        theme.selection_foreground = fg.clone();
    }
    if let Some(c) = config.get("cursor-color") {
        theme.cursor = c.clone();
    }
    if let Some(c) = config.get("cursor-text") {
        theme.cursor_accent = c.clone();
    }
    if let Some(c) = config.get("selection-background") {
        theme.selection_background = c.clone();
    }
    if let Some(c) = config.get("selection-foreground") {
        theme.selection_foreground = c.clone();
    }

    theme
}

fn parse_ghostty_file_multivalue(content: &str) -> (HashMap<String, String>, HashMap<u8, String>) {
    let mut config = HashMap::new();
    let mut palette = HashMap::new();

    for line in content.lines() {
        let line = line.trim();
        if line.is_empty() || line.starts_with('#') {
            continue;
        }
        if let Some((key, value)) = line.split_once('=') {
            let key = key.trim();
            let value = value.trim();
            if key == "palette" {
                // value is like "0=#45475a"
                if let Some((idx_str, color)) = value.split_once('=') {
                    if let Ok(idx) = idx_str.trim().parse::<u8>() {
                        palette.insert(idx, color.trim().to_string());
                    }
                }
            } else {
                config.insert(key.to_string(), value.to_string());
            }
        }
    }

    (config, palette)
}

pub fn load_theme() -> Theme {
    let config_path = match ghostty_config_path() {
        Some(p) => p,
        None => return Theme::default(),
    };

    let config_content = match fs::read_to_string(&config_path) {
        Ok(c) => c,
        Err(_) => return Theme::default(),
    };

    let (config, _palette) = parse_ghostty_file_multivalue(&config_content);

    // If a theme name is specified, load the theme file
    if let Some(theme_name) = config.get("theme") {
        if let Some(themes_dir) = ghostty_themes_dir() {
            let theme_path = themes_dir.join(theme_name);
            if let Ok(theme_content) = fs::read_to_string(&theme_path) {
                let (mut theme_config, theme_palette) =
                    parse_ghostty_file_multivalue(&theme_content);

                // Config file overrides theme file for explicit settings
                // But only for non-theme, non-palette keys that the user explicitly set
                for (key, value) in &config {
                    if key != "theme" {
                        theme_config.insert(key.clone(), value.clone());
                    }
                }

                return build_theme(&theme_config, &theme_palette);
            }
        }
    }

    // No theme reference, parse config directly
    let (config, palette) = parse_ghostty_file_multivalue(&config_content);
    build_theme(&config, &palette)
}

fn build_theme(config: &HashMap<String, String>, palette: &HashMap<u8, String>) -> Theme {
    let mut theme = load_theme_from_config(config);

    let palette_map: [(u8, &mut String); 16] = [
        (0, &mut theme.black),
        (1, &mut theme.red),
        (2, &mut theme.green),
        (3, &mut theme.yellow),
        (4, &mut theme.blue),
        (5, &mut theme.magenta),
        (6, &mut theme.cyan),
        (7, &mut theme.white),
        (8, &mut theme.bright_black),
        (9, &mut theme.bright_red),
        (10, &mut theme.bright_green),
        (11, &mut theme.bright_yellow),
        (12, &mut theme.bright_blue),
        (13, &mut theme.bright_magenta),
        (14, &mut theme.bright_cyan),
        (15, &mut theme.bright_white),
    ];

    for (idx, target) in palette_map {
        if let Some(color) = palette.get(&idx) {
            *target = color.clone();
        }
    }

    theme
}

#[tauri::command]
pub fn get_theme() -> Theme {
    load_theme()
}

File: electron/theme.ts
import fs from "node:fs";
import path from "node:path";
import os from "node:os";

export interface Theme {
  background: string;
  foreground: string;
  cursor: string;
  cursorAccent: string;
  selectionBackground: string;
  selectionForeground: string;
  black: string;
  red: string;
  green: string;
  yellow: string;
  blue: string;
  magenta: string;
  cyan: string;
  white: string;
  brightBlack: string;
  brightRed: string;
  brightGreen: string;
  brightYellow: string;
  brightBlue: string;
  brightMagenta: string;
  brightCyan: string;
  brightWhite: string;
}

const DEFAULT_THEME: Theme = {
  background: "#1e1e2e",
  foreground: "#cdd6f4",
  cursor: "#f5e0dc",
  cursorAccent: "#1e1e2e",
  selectionBackground: "#585b70",
  selectionForeground: "#cdd6f4",
  black: "#45475a",
  red: "#f38ba8",
  green: "#a6e3a1",
  yellow: "#f9e2af",
  blue: "#89b4fa",
  magenta: "#f5c2e7",
  cyan: "#94e2d5",
  white: "#a6adc8",
  brightBlack: "#585b70",
  brightRed: "#f37799",
  brightGreen: "#89d88b",
  brightYellow: "#ebd391",
  brightBlue: "#74a8fc",
  brightMagenta: "#f2aede",
  brightCyan: "#6bd7ca",
  brightWhite: "#bac2de",
};

function ghosttyConfigPath(): string | null {
  // macOS: ~/Library/Application Support/com.mitchellh.ghostty/config
  const macPath = path.join(
    os.homedir(),
    "Library",
    "Application Support",
    "com.mitchellh.ghostty",
    "config",
  );
  if (fs.existsSync(macPath)) return macPath;

  // XDG: ~/.config/ghostty/config
  const xdgPath = path.join(os.homedir(), ".config", "ghostty", "config");
  if (fs.existsSync(xdgPath)) return xdgPath;

  return null;
}

function ghosttyThemesDir(): string | null {
  const appThemes =
    "/Applications/Ghostty.app/Contents/Resources/ghostty/themes";
  if (fs.existsSync(appThemes)) return appThemes;

  const xdgThemes = path.join(
    os.homedir(),
    "Library",
    "Application Support",
    "ghostty",
    "themes",
  );
  if (fs.existsSync(xdgThemes)) return xdgThemes;

  return null;
}

function parseGhosttyFile(content: string): {
  config: Map<string, string>;
  palette: Map<number, string>;
} {
  const config = new Map<string, string>();
  const palette = new Map<number, string>();

  for (const line of content.split("\n")) {
    const trimmed = line.trim();
    if (!trimmed || trimmed.startsWith("#")) continue;
    const eqIdx = trimmed.indexOf("=");
    if (eqIdx < 0) continue;
    const key = trimmed.slice(0, eqIdx).trim();
    const value = trimmed.slice(eqIdx + 1).trim();

    if (key === "palette") {
      const palEq = value.indexOf("=");
      if (palEq >= 0) {
        const idx = parseInt(value.slice(0, palEq).trim(), 10);
        if (!isNaN(idx)) {
          palette.set(idx, value.slice(palEq + 1).trim());
        }
      }
    } else {
      config.set(key, value);
    }
  }

  return { config, palette };
}

function loadThemeFromConfig(config: Map<string, string>): Theme {
  const theme = { ...DEFAULT_THEME };

  const bg = config.get("background");
  if (bg) {
    theme.background = bg;
    theme.cursorAccent = bg;
  }
  const fg = config.get("foreground");
  if (fg) {
    theme.foreground = fg;
    theme.selectionForeground = fg;
  }
  if (config.has("cursor-color")) theme.cursor = config.get("cursor-color")!;
  if (config.has("cursor-text"))
    theme.cursorAccent = config.get("cursor-text")!;
  if (config.has("selection-background"))
    theme.selectionBackground = config.get("selection-background")!;
  if (config.has("selection-foreground"))
    theme.selectionForeground = config.get("selection-foreground")!;

  return theme;
}

function buildTheme(
  config: Map<string, string>,
  palette: Map<number, string>,
): Theme {
  const theme = loadThemeFromConfig(config);

  const paletteMap: [number, keyof Theme][] = [
    [0, "black"],
    [1, "red"],
    [2, "green"],
    [3, "yellow"],
    [4, "blue"],
    [5, "magenta"],
    [6, "cyan"],
    [7, "white"],
    [8, "brightBlack"],
    [9, "brightRed"],
    [10, "brightGreen"],
    [11, "brightYellow"],
    [12, "brightBlue"],
    [13, "brightMagenta"],
    [14, "brightCyan"],
    [15, "brightWhite"],
  ];

  for (const [idx, key] of paletteMap) {
    const color = palette.get(idx);
    if (color) {
      theme[key] = color;
    }
  }

  return theme;
}

const SETTINGS_PATH = path.join(
  os.homedir(),
  "Library",
  "Application Support",
  "Manor",
  "settings.json",
);

function loadSettings(): { themeName?: string } {
  try {
    return JSON.parse(fs.readFileSync(SETTINGS_PATH, "utf-8"));
  } catch {
    return {};
  }
}

function saveSettings(settings: { themeName?: string }): void {
  fs.mkdirSync(path.dirname(SETTINGS_PATH), { recursive: true });
  fs.writeFileSync(SETTINGS_PATH, JSON.stringify(settings, null, 2));
}

export type ThemeColors = Pick<
  Theme,
  | "red"
  | "green"
  | "yellow"
  | "blue"
  | "magenta"
  | "cyan"
  | "background"
  | "foreground"
>;

export class ThemeManager {
  private themeColorsCache: Record<string, ThemeColors> | null = null;

  /** Load a specific Ghostty theme file by name */
  loadGhosttyTheme(name: string): Theme | null {
    const themesDir = ghosttyThemesDir();
    if (!themesDir) return null;
    try {
      const content = fs.readFileSync(path.join(themesDir, name), "utf-8");
      const { config, palette } = parseGhosttyFile(content);
      return buildTheme(config, palette);
    } catch {
      return null;
    }
  }

  /** Load theme from Ghostty config (match Ghostty) */
  loadGhosttyConfigTheme(): Theme {
    const configPath = ghosttyConfigPath();
    if (!configPath) return DEFAULT_THEME;

    let content: string;
    try {
      content = fs.readFileSync(configPath, "utf-8");
    } catch {
      return DEFAULT_THEME;
    }

    const { config, palette: _palette } = parseGhosttyFile(content);

    const themeName = config.get("theme");
    if (themeName) {
      const themesDir = ghosttyThemesDir();
      if (themesDir) {
        try {
          const themeContent = fs.readFileSync(
            path.join(themesDir, themeName),
            "utf-8",
          );
          const { config: themeConfig, palette: themePalette } =
            parseGhosttyFile(themeContent);
          for (const [key, value] of config) {
            if (key !== "theme") themeConfig.set(key, value);
          }
          return buildTheme(themeConfig, themePalette);
        } catch {
          // Fall through
        }
      }
    }

    const { config: directConfig, palette } = parseGhosttyFile(content);
    return buildTheme(directConfig, palette);
  }

  /** Resolve a theme by name, handling special values __ghostty__ and __default__ */
  getThemeByName(name: string): Theme {
    if (!name || name === "__ghostty__") {
      return this.loadGhosttyConfigTheme();
    }
    if (name === "__default__") {
      return DEFAULT_THEME;
    }
    return this.loadGhosttyTheme(name) ?? DEFAULT_THEME;
  }

  /** Get the active theme based on saved settings */
  getTheme(): Theme {
    const settings = loadSettings();
    const themeName = settings.themeName;
    return this.getThemeByName(themeName ?? "__ghostty__");
  }

  /** Load color palettes for all themes (cached after first call) */
  async loadAllThemeColors(): Promise<Record<string, ThemeColors>> {
    if (this.themeColorsCache) return this.themeColorsCache;

    const themesDir = ghosttyThemesDir();
    if (!themesDir) return {};

    const result: Record<string, ThemeColors> = {};
    try {
      const files = (await fs.promises.readdir(themesDir)).filter(
        (f) => !f.startsWith("."),
      );
      const reads = files.map(async (name) => {
        try {
          const content = await fs.promises.readFile(
            path.join(themesDir, name),
            "utf-8",
          );
          const { config, palette } = parseGhosttyFile(content);
          const theme = buildTheme(config, palette);
          result[name] = {
            red: theme.red,
            green: theme.green,
            yellow: theme.yellow,
            blue: theme.blue,
            magenta: theme.magenta,
            cyan: theme.cyan,
            background: theme.background,
            foreground: theme.foreground,
          };
        } catch {
          // skip unreadable themes
        }
      });
      await Promise.all(reads);
    } catch {
      // ignore
    }
    this.themeColorsCache = result;
    return result;
  }

  /** Check if Ghostty config exists on this system */
  hasGhosttyConfig(): boolean {
    return ghosttyConfigPath() !== null;
  }

  /** Get the saved theme name from settings */
  getSelectedThemeName(): string {
    const settings = loadSettings();
    return settings.themeName ?? "__ghostty__";
  }

  /** Save the selected theme name */
  setSelectedThemeName(name: string): void {
    const settings = loadSettings();
    settings.themeName = name;
    saveSettings(settings);
  }
}
